src/envoy/auth/config.h

```diff
@@ -13,115 +13,74 @@
-        cb_(cb) {}
-  // AsyncClient::Callbacks
-  void onSuccess(MessagePtr &&response);
-  void onFailure(AsyncClient::FailureReason);
-
-  void Call(const std::string &uri);
-  void Cancel();
+// Struct to hold an issuer's config.
+struct IssuerInfo {
+  // URI for public key server
+  std::string uri;
+  // Envoy cluster name for public key server
+  std::string cluster;
 
- private:
-  Upstream::ClusterManager &cm_;
-  Upstream::ClusterInfoConstSharedPtr cluster_;
-  Optional<std::chrono::milliseconds> timeout_;
-  std::function<void(bool, const std::string &)> cb_;
-  AsyncClient::Request *request_;
-};
+  // Issuer name. e.g. "https://accounts.example.com"
+  std::string name;
 
-// Struct to hold an issuer's information.
-struct IssuerInfo : public Logger::Loggable<Logger::Id::http> {
-  // This constructor loads config from JSON. When public key is given via URI,
-  // it just keeps URI and cluster name, and public key will be fetched later,
-  // namely in decodeHeaders() of the filter class.
-  IssuerInfo(Json::Object *json);
-  // Allow default construct.
-  IssuerInfo(const std::string &name, const std::string &uri,
-             const std::string &cluster, Pubkeys::Type pkey_type,
-             std::vector<std::string> &&audiences)
-      : uri_(uri),
-        cluster_(cluster),
-        name_(name),
-        pkey_type_(pkey_type),
-        audiences_(std::move(audiences)) {}
-
-  // True if the config loading in the constructor or fetching public key failed
-  bool failed_ = false;
-
-  // True if the public key is loaded
-  bool loaded_ = false;
-
-  std::string uri_;      // URI for public key
-  std::string cluster_;  // Envoy cluster name for public key
-
-  std::string name_;               // e.g. "https://accounts.example.com"
-  Pubkeys::Type pkey_type_;        // Format of public key.
-  std::unique_ptr<Pubkeys> pkey_;  // Public key
+  // Format of public key.
+  Pubkeys::Type pubkey_type;
+  // public key value. Public key can be specified in the Envoy config.
+  std::string pubkey_value;
 
+  // Time to expire a cached public key (sec).
+  // 0 means never expired.
+  int64_t pubkey_cache_expiration_sec{};
+
+  // The audiences should be matched from JWT.
+  std::set<std::string> audiences;
+
+  // Check if an audience is allowed.
   // If audiences is an empty array or not specified, any "aud" claim will be
   // accepted.
-  std::vector<std::string> audiences_;
-  bool IsAudienceAllowed(const std::string &aud);
+  bool IsAudienceAllowed(const std::string &aud) const {
+    return audiences.empty() || audiences.find(aud) != audiences.end();
+  }
+
+  // Validate the issuer config.
+  // Return error message if invalid, otherwise return empty string.
+  std::string Validate() const;
 };
 
-// A config for Jwt auth filter
-struct JwtAuthConfig : public Logger::Loggable<Logger::Id::http> {
-  // Load the config from envoy config.
+// A config for jwt-auth filter
+class Config : public Logger::Loggable<Logger::Id::http> {
```

Comment:
Any reason to change the class name? I like the old name (or alternate option JwtFilterConfig), which give me some clue where it come from. In this PR alone there are already bunch of things call "config" and this rename would make it harder to follow.

Reply:
+1, the namespace is neither clear, so let's keep JwtAuthConfig.